crates/runtime-table-partition/src/insert.rs

```diff
@@ -0,0 +1,318 @@
+        if children.len() != 1 {
+            return Err(DataFusionError::Plan(
+                "PartitionInsertExec requires exactly one child".to_string(),
+            ));
+        }
+        Ok(Arc::new(Self::new(
+            Arc::clone(&children[0]),
+            self.partition_by.clone(),
+            Arc::clone(&self.creator),
+            Arc::clone(&self.partitions),
+            self.insert_op,
+            Arc::clone(&self.schema),
+        )))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream, DataFusionError> {
+        // We want to stream the input RecordBatches, partition them into
+        // RecordBatches destined for a particular partitioned file, stream
+        // them into that partitioned file without buffering all of the data,
+        // and call `insert_into` on the DuckDB Table Providers only once
+        // because data are rewritten every call to `insert_into`.
+        //
+        // We also have an interesting mix of async/sync contexts here in
+        // `execute` which we deal with by using `block_in_place` and
+        // `Handle::block_on`.
+        //
+        // In one task, we stream the input RecordBatches and partition them
+        // according to the partition_by expression. If the partition has not
+        // yet been created, we create it. We also create a channel for sending
+        // RecordBatches. We make a LazyMemExec ExecutionPlan which we pass to
+        // that partition's `insert_into` one time. The LazyMemExec is passed
+        // a generator that we create which generates record batches out of the
+        // receiving end of the partition's RecordBatch channel. Each partition
+        // `insert_into` invokation is ran in a separate task. So we have one
+        // partitioning task and N insertion tasks where N is the number of
+        // partitions.
+
+        let session_config = context.session_config();
+        let ctx = SessionContext::new_with_config(session_config.clone());
+
+        if partition != 0 {
+            return Err(DataFusionError::Execution(
+                "PartitionInsertExec only supports single partition".to_string(),
+            ));
+        }
+
+        let insert_op = self.insert_op;
+        let df_schema = DFSchema::try_from(Arc::clone(&self.schema))?;
+        let props = ExecutionProps::new();
+        let physical_expr = create_physical_expr(&self.partition_by, &df_schema, &props)?;
+        let schema = Arc::clone(&self.schema);
+        let creator = Arc::clone(&self.creator);
+        let partition_providers = Arc::clone(&self.partitions);
+        let input = Arc::clone(&self.input);
+        let task_ctx = Arc::clone(&context);
+
+        let mut partition_senders: HashMap<String, Sender<RecordBatch>> = HashMap::new();
+
+        let mut stream = execute_stream(input, task_ctx)?;
+
+        block_in_place(move || {
```

Comment:
We shouldn't do this, we should do any async work in the stream that is returned.